### ISSUE TITLE
Signature verification fails for nearly every attestations

### ISSUE THREAD

**Issue by nisdas** [prysmaticlabs/prysm#3083]
```
[2019-07-26 23:21:48]  INFO rpc: Deleting failed pending attestation from DB error=could not verify indexed attestation: attestation aggregation signature did not verify headRoot=0xebc601c3be2d75b3f0633ebf slot=49
```

Although this attestation is invalid as it is > 2 epochs old, signature verification should not fail for it, as it has a valid signature

**Comment by terencechain** [comment]
Do we roll back state to the the > 2 epochs old snap shot for verifying these old signatures?

**Comment by nisdas** [comment]
@terencechain I dont think so, might be a reason why it fails. But even then why does the state influence signature verification ? we just verify that the root of the indexed attestation has been signed correctly.

**Comment by prestonvanloon** [comment]
We are facing similar issues without the start shard cache. Almost every attestation fails to verify in an epoch without `--enable-start-shard-cache`.

**Comment by prestonvanloon** [comment]
We've observed that StartShard is changing with a different states and same epoch. This is a misuse of the caching because the cache was not updated as the epoch is the cache key.

This change in start shard on epoch transition is causing assignments to be invalid / outdated. One solution for this is to fetch assignments at every slot and the problem no longer exists.

**Comment by prestonvanloon** [comment]
Fetching assignments on every slot still fails for this issue on epoch start slot or first slot thereafter.

**Comment by terencechain** [comment]
More findings...
So here's the shard assignments. Top is beacon chain. Bottom is validator client:
```
[2019-07-30 15:46:05]  WARN blockchain: index 0, pubkey 84c3f8c975e9, slot 21, shard 3, isProposer false, committee [4 57 18 6 49 0 3 48]
[2019-07-30 15:46:05]  WARN blockchain: index 1, pubkey 8b977879d6c8, slot 17, shard 7, isProposer false, committee [63 44 23 47 20 22 1 58]
[2019-07-30 15:46:05]  WARN blockchain: index 19, pubkey b4c22517bff5, slot 19, shard 1, isProposer false, committee [52 7 46 15 19 5 37 21]
```
```
[2019-07-30 15:46:04]  INFO validator: New assignment attesterSlot=21 committee=[4 57 18 6 49 0 3 48] proposerSlot=Not proposing pubKey=84c3f8c975e9 shard=4 status=ACTIVE
[2019-07-30 15:46:04]  INFO validator: New assignment attesterSlot=17 committee=[63 44 23 47 20 22 1 58] proposerSlot=Not proposing pubKey=8b977879d6c8 shard=0 status=ACTIVE
[2019-07-30 15:46:04]  INFO validator: New assignment attesterSlot=19 committee=[52 7 46 15 19 5 37 21] proposerSlot=Not proposing pubKey=b4c22517bff5 shard=2 status=ACTIVE
```
shards 3, 7, 1
shards 4, 0 ,2
all validators have the wrong assignments

**Comment by nisdas** [comment]
This is resolved by #3112
